bazel/linux/test.bzl

```diff
@@ -4,6 +4,83 @@ load("//bazel/utils:macro.bzl", "mconfig", "mcreate_rule")
 load("//bazel/utils:exec_test.bzl", "exec_test")
 load("//bazel/linux:bundles.bzl", "kunit_bundle")
 
+def _test_runner(ctx):
+    script_begin = """#!/bin/bash
+set -eu -o pipefail
+GREEN="\033[32m"
+RED="\033[31m"
+RESET="\033[0m"
+failed_outputs=()
+tests=0
+failed=0
+run_test() {
+    local title="$1"
+    local script="$2"
+    local success=0
+    local output
+    local status
+    local start_time_ns
+    local stop_time_ns
+    local duration_ns
+
+    printf "%-73s" "${title}"
+    start_time_ns=$(($(date +%s%N)))
+
+    output=$("${script}" 2>&1) && success=1
+
+    tests=$[tests + 1]
+    if [ $success -eq 1 ]; then
+        status="${GREEN}PASSED${RESET}"
+    else
+        failed=$[failed + 1]
+        status="${RED}FAILED${RESET}"
+        failed_outputs+="==================== Test output for ${title}:\\n${output}\\n"
+    fi
+    stop_time_ns=$(($(date +%s%N)))
+    duration_ns=$(($stop_time_ns - $start_time_ns))
+    printf "%s in %.1fs\\n" "$status" "$duration_ns""e-9"
+}
+"""
+
+    script_end = """
+for o in "${failed_outputs[@]}"; do
+    echo -ne "$o"
+done
+if [ $failed -ne 0 ]; then
+    echo "================================================================================"
+fi
+succeeded=$[tests - failed]
+echo "Executed ${tests} test(s): ${succeeded} test(s) pass(es) and ${failed} fail(s)."
+exit $failed
+"""
+
+    tests = []
+    runfiles = ctx.runfiles()
+    for test in ctx.attr.tests:
+        binary = test.files_to_run.executable
+        tests.append("run_test {title} {script}".format(
+            title = test.label,
+            script = binary.short_path,
+        ))
+        runfiles = runfiles.merge(ctx.runfiles([binary]))
+
+    script = ctx.actions.declare_file("{}_test_runner.sh".format(ctx.attr.name))
+    ctx.actions.write(script, script_begin + "\n".join(tests) + script_end)
+    print(script_begin + "\n".join(tests) + script_end)
+    return [DefaultInfo(runfiles = runfiles, executable = script)]
```

Comment:
so... the `tests` labels passed here... if they are `vm_bundle()` targets, this rule does not propagate the `pre` step and `check` step to the runner below. So a `vm_bundle` passed to a qemu_test target would silently break?

Reply:
(it may be ok to do so, but I would be a bit more defensive here, see other comment above when parsing list of targets)

Reply:
This is pending. I will work on it and update the PR.

Reply:
I've created a 2 refactoring PRs and updated this PR with support for `vm_bundle`.